google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/BlockingPullSubscriberImpl.java

```diff
@@ -0,0 +1,126 @@
+import java.util.Collection;
+import java.util.Deque;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+public class BlockingPullSubscriberImpl implements BlockingPullSubscriber {
+
+  private final Subscriber underlying;
+
+  @GuardedBy("this")
+  private Optional<CheckedApiException> error = Optional.empty();
+
+  @GuardedBy("this")
+  private Deque<SequencedMessage> messages = new ArrayDeque<>();
+
+  @GuardedBy("this")
+  private Optional<SettableApiFuture<Void>> notification = Optional.empty();
+
+  public BlockingPullSubscriberImpl(
+      SubscriberFactory factory, FlowControlSettings settings, SeekRequest initialSeek)
+      throws CheckedApiException {
+    underlying = factory.newSubscriber(this::addMessages);
+    underlying.addListener(
+        new Listener() {
+          @Override
+          public void failed(State state, Throwable throwable) {
+            fail(ExtractStatus.toCanonical(throwable));
+          }
+        },
+        MoreExecutors.directExecutor());
+    underlying.startAsync().awaitRunning();
+    try {
+      underlying.seek(initialSeek).get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw ExtractStatus.toCanonical(e);
+    }
+    underlying.allowFlow(
+        FlowControlRequest.newBuilder()
+            .setAllowedMessages(settings.messagesOutstanding())
+            .setAllowedBytes(settings.bytesOutstanding())
+            .build());
+  }
+
+  private synchronized void fail(CheckedApiException e) {
+    error = Optional.of(e);
+    if (notification.isPresent()) {
+      notification.get().setException(e);
+      notification = Optional.empty();
+    }
+  }
+
+  private synchronized void addMessages(Collection<SequencedMessage> new_messages) {
+    messages.addAll(new_messages);
+    if (notification.isPresent()) {
+      notification.get().set(null);
+      notification = Optional.empty();
+    }
+  }
+
+  @Override
+  public synchronized Future<Void> onData() {
+    if (notification.isPresent()) {
+      notification
+          .get()
+          .setException(new InterruptedException("Interruped and superseded by newer onData call"));
+      notification = Optional.empty();
+    }
+    if (error.isPresent()) {
+      return ApiFutures.immediateFailedFuture(error.get());
+    }
+    if (!messages.isEmpty()) {
+      return ApiFutures.immediateFuture(null);
+    }
+    notification = Optional.of(SettableApiFuture.create());
+    return notification.get();
+  }
+
+  @Override
+  public synchronized Optional<SequencedMessage> messageIfAvailable() throws CheckedApiException {
+    if (error.isPresent()) {
+      throw error.get();
+    }
+    if (!messages.isEmpty()) {
+      return Optional.of(Objects.requireNonNull(messages.pollFirst()));
+    }
+    return Optional.empty();
+  }
+
+  @Override
+  public void close() {
+    underlying.stopAsync().awaitTerminated();
```

Comment:
nit. probably worth making this fail future calls by making this method set error to UNAVAILABLE with "client shut down" before calling stopAsync

```
synchronized {
  if (error.isPresent()) return;
  error = Optional.of(...);
}
underlying.stopAsync().awaitTerminated();
```

since you don't want to call awaitTerminated in a synchronized block.